[Breaking change] Remove all use of the language service. Use only the Program.

andy-ms opened the pull request:
#### PR checklist

- [ ] Addresses an existing issue: #1969
  - Should fix #2181 (that error was in document highlights, which we no longer call)
  - Fixes #2210
- [X] New feature, bugfix, or enhancement
  - [ ] Includes tests
- [ ] Documentation update

#### What changes did you make?

Tslint now only provides the `Program`, not the `LanguageService`. `LanguageService` is for editors; tslint should use the `Program` and `TypeChecker` API.
This would be break any custom typed rules, so should be in a major version change. So long as the custom typed rules do not use the language service, conversion is simple. See [await-promise](https://github.com/palantir/tslint/pull/2235/files#diff-f598bb4bf89a3fbf85ddc367281087f2) for an example.

Only `no-use-before-declare` and `no-unused-variable` used the language service.
Rewrote `no-use-before-declare`. It now gets the symbol at each use and checks that it comes after the declaration.
I think `no-unused-variable` was going to be deprecated anyway, so removed it. It would be possible to rewrite that too if you want to keep it.

As a side-effect, fixes #1969 since we don't have `wrapProgram`, which creates a new language service, which then has its own program. This probably helps performance too, but I haven't tested.

Also changed tslint's own linting to use the program. Using typed rules in this repo could be done in another PR.

Will rebase after #2234 is in.

ajafff commented on src/language/rule/typedRule.ts:
```diff
@@ -31,5 +31,5 @@ export abstract class TypedRule extends AbstractRule {
     }
 
-    public abstract applyWithProgram(sourceFile: ts.SourceFile, languageService: ts.LanguageService): RuleFailure[];
+    public abstract applyWithProgram(sourceFile: ts.SourceFile, languageService: ts.Program): RuleFailure[];
```
Also rename the parameter

nchen63 commented:
actually, #1481 is no longer being deprecated due to community response

nchen63 commented:
The parameter originally used `program`, and was changed to use language service in #1608 

Need to verify that the original issue is doesn't break

andy-ms commented:
Rewrote `no-unused-variable` to not need the language service. Done by creating a duplicate program with the unused-checking compiler options set. Added workaround for Microsoft/TypeScript#9944 that doesn't require `checkEdit`.
This rule would presumably be unnecessary if we fixed that bug and implemented Microsoft/TypeScript#11051.

andy-ms commented:
Note that this makes `no-unused-variable` and `no-use-before-declare` typed rules, while before they worked without a type checker.

nchen63 commented:
That might be a problem for people who use plug-ins like the VSCode-TSLint since typed rules are not supported. People seem to use this rule to see the warnings instead of using the compiler flag so it doesn't block compilation; see https://github.com/palantir/tslint/issues/1481#issuecomment-262498483.

ajafff commented:
I'm already working on a solution to use both rules without type checking.
`no-use-before-declare` is no problem.
`no-unused-variable` would only find unused private class members with type checking enabled. The problem with implicit required imports may also still be there - haven't looked at the code yet - so we would exclude them from the check without type checking. This way we would at least get most of the failures and find the rest with type checking.

andy-ms commented:
It may be possible to create a program consisting of a single file and run the rest as usual.
